File: syntax-highlighting/get-GP-functions.pl
#!/usr/bin/perl -T
use strict;
use warnings;

use Capture::Tiny 'capture';
use Array::Utils qw(:all);
use Regexp::Assemble;	# only needed if writing regex

# Set these to the locations of these files on your system.
my $src = '../src';
my $whatnow = "$src/gp/whatnow.h";
my $paridesc = "$src/desc/pari.desc";
my $gp2c = '/usr/local/bin/gp2c';


# Get raw output from gp2c.
sub gp2c {
	$ENV{PATH} = "/bin:/usr/bin";	# Otherwise taint mode won't let system run
	my ($stdout, $stderr, $exit) = capture {
		system( $gp2c, ('-l') );
	};
	die if $exit != 0 || $stderr ne '';
	return split("\n", $stdout);
}


# Get a list of functions from gp2c.
sub getFunctions {
	my @func = ();
	my @operator = ();	# currently ignored
	for my $func (gp2c()) {
		if ($func =~ /^([^ ]+) ([^ ]+) (.*)$/i) {
			my $GP = $1;
			my $PARI = $2;	# currently ignored
			my $prototype = $3;	# currently ignored
			if ($GP =~ /^[a-z][a-z0-9_]*$/i) {
				push @func, $GP;
			} else {	
				push @operator, $GP;
			}
		} elsif ($func =~ /^[\s\r\n]*$/) {
		} else {
			print "Unrecognized gp2c line: '$func'\n";
		}
	}
	return sort(@func);
}


sub getRemoved {
	my @removed = ();
	open my $in, '<', $whatnow or die "Could not find src/gp/whatnow.h";
	while(my $line = <$in>) {
		if ($line =~ /^\{"[^" ]+", *_SAME\}/) {
		} elsif ($line =~ /^\{"([a-zA-Z][a-zA-Z0-9_]*)", *_REMOV\}/) {
			push @removed, $1;	# Explicitly marked as removed
		} elsif ($line =~ /\{"([a-zA-Z][a-zA-Z0-9_]*)","([^" ]+)","[^"]*","[^"]*"\}/) {
			my $oldname = $1;
			my $newname = $2;
			if ($oldname eq $newname) {
				# No need to do anything -- only the arguments changed
			} else {
				push @removed, $oldname;
			}
		}
	}
	close $in;
	return sort(@removed);
}


sub getObsolete {
	my @obsolete = ();
	my $name = '';
	open my $in, '<', $paridesc or die "Could not find src/desc/pari.desc";
	while(my $line = <$in>) {
		if ($line =~ /^Function: ([a-zA-Z][a-zA-Z0-9_]*)$/) {	# Function
			$name = $1;
		} elsif ($line =~ /^Function: ([^a-zA-Z]+)$/) {	# Operator
			$name = $1;
		} elsif ($line =~ /^Function: (_[a-zA-Z0-9_]+)$/) {	# gp2c internal function
			$name = $1;
		} elsif ($line =~ /^Function: (_\.[a-zA-Z][a-zA-Z0-9_]*)$/) {	# member function
			$name = $1;
		} elsif ($line =~ /^Function:/) {	# weird, skipped
			# Currently this includes only 'O(_^_)' and '_^s'
			$name = '';
		} elsif ($line =~ /^Obsolete: (.+)$/) {
			my $obsoleteDate = $1;	# currently ignored
			push @obsolete, $name if $name ne '';
		}
	}
	return sort(@obsolete);
}


my @func = getFunctions();
my @removed = getRemoved();
my @obsolete = getObsolete();

# If it's in both, one function was removed and another with the same name added.
@removed = array_minus(@removed, @func);

my @bad = sort(unique(@removed, @obsolete));
my @good = sort(array_minus(@func, @obsolete));

print 'Good functions: ' . join(', ', @good) . "\n";
print 'Bad functions: ' . join(', ', @bad) . "\n";
print 'In total there are ' . scalar(@good) . ' good and ' . scalar(@bad) . " bad functions.\n";

#my $ra = Regexp::Assemble->new;
#$ra->add(@func);
#my $r = $ra->re;
#$r =~ s/\(\?:/\(/g;
#print "$r\n";

File: syntax-highlighting/get-GP-functions.py
#!/usr/bin/env python3

import re
import subprocess
from typing import List
from array import array
import argparse

# Set these to the locations of these files on your system.
src = '../../pari/src/' # '../src'
whatnow = f'{src}/gp/whatnow.h'
paridesc = f'{src}/desc/pari.desc'
gp2c = '/usr/local/bin/gp2c'

parser = argparse.ArgumentParser(description='Show GP functions')
parser.add_argument('--format', choices=['1', 'comma'], default='1', help='1 (one function per line), comma (comma-separated list)')
parser.add_argument('--show', choices=['good', 'bad', 'removed', 'obsolete', 'current', 'all'], default='good', help='good (current functions which are not obsolete), obsolete (current functions which are slated to be removed), removed (functions no longer available), bad (removed or obsolete functions), current (all available functions, whether obsolete or not), all (all functions, including those removed)')
args = parser.parse_args()

def array_minus(a: list, b: list) -> list:
    return list(set(a) - set(b))

def unique(a: list) -> list:
    return list(set(a))


# Get raw output from gp2c.
def run_gp2c() -> List[str]:
    env = {'PATH': '/bin:/usr/bin'}
    proc = subprocess.run([gp2c, '-l'], env=env, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
    return proc.stdout.decode().splitlines()


# Get a list of functions from gp2c.
def get_functions() -> List[str]:
    func = []
    operator = []  # currently ignored
    for line in run_gp2c():
        if match := re.match(r'^([^ ]+) ([^ ]+) (.*)$', line, re.IGNORECASE):
            GP, PARI, prototype = match.groups()
            if re.match(r'^[a-z][a-z0-9_]*$', GP, re.IGNORECASE):
                func.append(GP)
            else:
                operator.append(GP)
        elif re.match(r'^[\s\r\n]*$', line):
            pass
        else:
            print(f"Unrecognized gp2c line: '{line}'", file=sys.stderr)
    return sorted(func)


def get_removed() -> List[str]:
    removed = []
    with open(whatnow, 'r') as f:
        for line in f:
            if match := re.match(r'^\{"[^" ]+", *_SAME\}', line):
                pass
            elif match := re.match(r'^\{"([a-zA-Z][a-zA-Z0-9_]*)", *_REMOV\}', line):
                removed.append(match.group(1))  # Explicitly marked as removed
            elif match := re.match(r'\{"([a-zA-Z][a-zA-Z0-9_]*)","([^" ]+)","[^"]*","[^"]*"\}', line):
                oldname, newname = match.groups()
                if oldname != newname:
                    removed.append(oldname)
    return sorted(removed)


def get_obsolete() -> List[str]:
    obsolete = []
    name = ''
    with open(paridesc, 'r') as f:
        for line in f:
            if match := re.match(r'^Function: ([a-zA-Z][a-zA-Z0-9_]*)$', line):
                name = match.group(1)  # Function
            elif match := re.match(r'^Function: ([^a-zA-Z]+)$', line):
                name = match.group(1)  # Operator
            elif match := re.match(r'^Function: (_[a-zA-Z0-9_]+)$', line):
                name = match.group(1)  # gp2c internal function
            elif match := re.match(r'^Function: (_\.[a-zA-Z][a-zA-Z0-9_]*)$', line):
                name = match.group(1)  # member function
            elif re.match(r'^Function:', line):  # weird, skipped
                # Currently this includes only 'O(_^_)' and '_^s'
                name = ''
            elif match := re.match(r'^Obsolete: (.+)$', line):
                obsoleteDate = match.group(1)  # currently ignored
                if name != '':
                    obsolete.append(name)
    return sorted(obsolete)


def show(out, fmt):
    if fmt == '1':
       print('\n'.join(out))
    elif fmt == 'comma':
        print(', '.join(out))

func = get_functions()
removed = get_removed()
obsolete = get_obsolete()

# If it's in both, one function was removed and another with the same name added.
removed = list(set(removed) - set(func))

bad = sorted(list(set(removed) | set(obsolete)))
good = sorted(list(set(func) - set(obsolete)))

if args.show == 'good':
    show(good, args.format)
elif args.show == 'bad':
    show(bad, args.format)
elif args.show == 'removed':
    show(removed, args.format)
elif args.show == 'obsolete':
    show(obsolete, args.format)
elif args.show == 'current':
    show(func, args.format)
elif args.show == 'all':
    if args.format == '1':
        all = sorted(list(set(good) | set(bad)))
        show(all, '1')
    else:
        for s in ['good', 'bad', 'removed', 'obsolete']:
            print(f"{s}:")
            show(s, args.format)

#print(f"In total there are {len(good)} good, {len(removed)} removed, and {len(obsolete)} obsolete functions.")
